src/VecSim/algorithms/brute_force/bf_batch_iterator.cpp

```diff
@@ -0,0 +1,139 @@
+            upperBound = TopCandidates.top().first;
+        } else {
+            if (scores[i].first >= upperBound) {
+                continue;
+            } else {
+                TopCandidates.emplace(scores[i].first, scores[i].second);
+                TopCandidatesIndices[scores[i].second] = i;
+                // remove the furthest vector from the candidates and from the label->index mappings
+                TopCandidatesIndices.erase(TopCandidates.top().second);
+                TopCandidates.pop();
+                upperBound = TopCandidates.top().first;
+            }
+        }
+    }
+
+    auto *results = array_new_len<VecSimQueryResult>(TopCandidates.size(), TopCandidates.size());
+    for (int i = (int)TopCandidates.size() - 1; i >= 0; --i) {
+        VecSimQueryResult_SetId(results[i], TopCandidates.top().second);
+        VecSimQueryResult_SetScore(results[i], TopCandidates.top().first);
+        // Invalidate vector score, so we won't return it again in the next iterations.
+        scores[TopCandidatesIndices[TopCandidates.top().second]].first = INVALID_SCORE;
+        TopCandidates.pop();
+    }
+    return results;
+}
+
+VecSimQueryResult *BF_BatchIterator::selectBasedSearch(size_t n_res) {
+    // We return up to n_res vectors, the remaining vectors size is an upper bound.
+    if (n_res > scores.size()) {
+        n_res = scores.size();
+    }
+    auto n_th_element_pos = scores.begin() + (int)n_res;
+    // This will perform an in-place partition of the elements in the array, based on the
+    // n-th element as the pivot - every element with a lower will be placed before it, and
+    // all the rest will be placed after.
+    std::nth_element(scores.begin(), n_th_element_pos, scores.end(), cmp);
+
+    auto *results = array_new<VecSimQueryResult>(n_res);
+    for (size_t i = 0; i < n_res; i++) {
+        if (isnan(scores[i].first)) {
+            continue;
+        }
+        results = array_append(results, VecSimQueryResult{});
+        VecSimQueryResult_SetId(results[array_len(results) - 1], scores[i].second);
+        VecSimQueryResult_SetScore(results[array_len(results) - 1], scores[i].first);
+    }
+    // Erase the returned results slice from the scores array
+    scores.erase(scores.begin(), n_th_element_pos);
+    return results;
+}
+
+BF_BatchIterator::BF_BatchIterator(const void *query_vector, const BruteForceIndex *bf_index)
+    : VecSimBatchIterator(query_vector), id(BF_BatchIterator::next_id), index(bf_index) {
+    BF_BatchIterator::next_id++;
+}
+
+VecSimQueryResult_List BF_BatchIterator::getNextResults(size_t n_res,
+                                                        VecSimQueryResult_Order order) {
+    assert((order == BY_ID || order == BY_SCORE) &&
+           "Possible order values are only 'BY_ID' or 'BY_SCORE'");
+    // Only in the first iteration we need to compute all the scores
+    if (scores.empty() && getResultsCount() == 0) {
+        scores.reserve(getIndex()->indexSize());
```

Comment:
please let's use the convention that everything that is owned/called by the local instance, will be called from `this`
so `this->getIndex()->indexSize()`.
Please maintain it throughout the file